Add Manual bbox entry option to search page

stephenkilbourn opened the pull request:
As requested in PR #302 , This is breaking out a new feature into its own PR.  

AreaSelect            |  BBoxEntry
:-------------------------:|:-------------------------:
<img width="684" alt="defaultAreaSelect" src="https://github.com/radiantearth/stac-browser/assets/7388976/87edaf01-3a22-465a-9ce1-5ca156bd0bfd">|<img width="705" alt="newBBoxEntry" src="https://github.com/radiantearth/stac-browser/assets/7388976/bf2482a5-4ead-4213-a88b-b60535fdc4a1">

This PR is intended to add a new BBEntry component that allows the entry of [`x_min, y_min, x_max, y_max]` instead of using a bounding box selection on the map. This new component should help accessibility and help make it clear the coordinates being searched by a user.

The default behavior on the search page is still to show the areaSelect Component if a user checks the `Filter by spatial extent` option.  However, now there is a radio button option to toggle to "Enter coordinates". This option hides the area select and allows a user to manually enter the coordinates.

The input boxes are limited to numbers ±180 or ±90 via the bootstrap input component. 

I did add two English terms that would need translations. If there is a way for me to request that via crowdin, then I can do that:
```
    "defineBbox": {
      "map": "Draw bounding box on map",
      "text": "Enter coordinates"
    },
```

m-mohr commented on src/components/SearchFilter.vue:
```diff
@@ -27,8 +27,22 @@
         <b-form-group v-if="canFilterExtents" :label="$t('search.spatialExtent')" :label-for="ids.bbox">
-          <b-form-checkbox :id="ids.bbox" v-model="provideBBox" value="1" @change="setBBox()">{{ $t('search.filterBySpatialExtent') }}</b-form-checkbox>
-          <Map class="mb-4" v-if="provideBBox" :stac="stac" selectBounds @bounds="setBBox" scrollWheelZoom />
+          <b-form-checkbox :id="ids.bbox" v-model="provideBBox" @change="setBBox()">{{ $t('search.filterBySpatialExtent') }}</b-form-checkbox>
```
The UI looks a bit like tabs, but are not. I think this should either normal radio checkboxes or proper tabs.

stephenkilbourn replied:
I like the tab suggestion.

m-mohr commented on src/components/BBoxEntry.vue:
```diff
@@ -0,0 +1,85 @@
+  <b-form-group>
+    <b-form-row>
+      <b-col>
+        <b-form-group label="x_min" label-for="x_min">
```
Not sure how this is usually done in forms, but x_min seems a little too technical as a user readable label.

stephenkilbourn replied:
I asked around and didn't find a clear answer. I could do something like Lat max/min and Lon max/min.  Would that be clearer to most users?

m-mohr replied:
Maybe follow the specification terms? I guess there are names hidden in the API spec documents.

stephenkilbourn replied:
looking at the geojson spec, the bounding box uses the terms "southwest" and "northeast" and avoids min/max to avoid confusion when crossing the antimeridian.  I've updated the labels to Southwest latitude/longitude and Northeast latitude/longitude to match that.

m-mohr commented:
Thanks for the PR. Is the bbox synced between map and form inputs?

stephenkilbourn commented:
The coordinates drawn with the area select on the map sync to the form inputs.  I.e. a user can draw a box, and then switch to the form input and continue to manually edit them.

Unless I missed something, the areaSelect doesn't accept a starting bounding box to visually set the handles.  Right now, switching from the form inputs back to the bounding box resets the query.bbox value back to the default, so a user would need to re-draw.

The different behaviors does feel strange.  I could make it reset any time a user changes tabs to be consistent. Ideal scenario might be updating the the areaSelect to display the bounds and show both manual and visual map entries at the same time with state synced instead of on separate tabs.

m-mohr commented:
I guess the area select tool needs some work. It's a very simple thing that I found on the internet, but I guess others have solved that better somewhere...

stephenkilbourn commented:
with the latest update, submitting the search will submit the bbox on whichever tab was last edited. i.e. if a user edits with the manual entry, then switches to the map tab without drawing there, the bbox used in the search is still the one from the manual text entry.  The area select still does not update what is displayed, though.

m-mohr commented on src/components/SearchFilter.vue:
```diff
@@ -30,18 +30,15 @@
+                  <Map class="mb-4" :stac="stac" selectBounds @bounds="setBBox" scrollWheelZoom />
+                </b-tab>
+                <b-tab v-model="bboxSelectionStyle" :title="$t('search.defineBbox.text')">
+                  <BBoxEntry :bbox="query.bbox || [-180, -80, 180, 80]" @updateBBoxArray="updateBBoxArray" />
```
Defaults should be specified in the component prop, not in the template.

stephenkilbourn replied:
done.  Because the `query.bbox` is null by default I now let the component handle null values. I also added validation that if passed an array, the array must be an array of 4 numbers.

m-mohr commented on src/components/SearchFilter.vue:
```diff
@@ -544,6 +542,9 @@ export default {
       let bbox = null;
       if (this.provideBBox) {
+        if(this.bboxSelectionStyle ==="text") {
+          console.log("bbox selection is text")
```
Try to avoid console.logs in the code.

m-mohr commented on src/components/BBoxEntry.vue:
```diff
@@ -83,3 +85,11 @@ props: {
 </script>
+
+<style lang="scss">
+@import "../theme/variables.scss";
```
If you don't use the variables, you don't need to import them.

m-mohr commented on src/components/BBoxEntry.vue:
```diff
@@ -68,28 +68,28 @@ import { BFormInput, BFormGroup} from 'bootstrap-vue';
+      type: [Array, null],
+      default: () => ([-180, -80, 180, 80]),
+      validator(value) {
+        return value.length === 4 && value.every((e) => typeof e === 'number');
```
```suggestion
        return Array.isArray(value) && value.length === 4 && value.every((e) => typeof e === 'number');
```

stephenkilbourn commented:
ok, I added a commit that gave the area select an setInitialBounds method.  That will set the area box height and width to the  same value manually entered on the manual entry tab

m-mohr commented on src/components/BBoxEntry.vue:
```diff
@@ -0,0 +1,95 @@
+  <b-container>
+    <b-form-row>
+      <b-col>
+        <b-form-group label="Southwest Latitude" label-for="sw_latitude">
```
The labels needs to be added to the locales.

m-mohr commented on src/components/map/leaflet-areaselect.js:
```diff
@@ -36,6 +38,20 @@ L.AreaSelect = L.Class.extend({
     return this;
   },
 
+  setInitialBounds: function ( bounds) {
```
I just found my old code from another tool again:

```json
    setBounds: function(bounds) {
        if (Array.isArray(bounds) && bounds.length >= 4) {
          bounds = L.LatLngBounds([bounds[0], bounds[1]], [bounds[2], bounds[3]]);
        }
        this.map.fitBounds(bounds);

        var bottomLeft = this.map.latLngToContainerPoint(bounds.getSouthWest());
        var topRight = this.map.latLngToContainerPoint(bounds.getNorthEast());

        this.setDimensions({
            width: Math.abs(bottomLeft.x - topRight.x),
            height: Math.abs(bottomLeft.y - topRight.y)
        });
    },
```

I think this might be a bit more universal and supports the Leaflet Bounds for consistency with getBounds. Would you mind updating it?

m-mohr commented on src/components/Map.vue:
```diff
@@ -161,6 +165,9 @@ export default {
     }
   },
   watch: {
+    bbox: function(newVal) {
```
What does this watcher do?

stephenkilbourn replied:
no longer needed, so removing

m-mohr commented on src/components/map/leaflet-areaselect.js:
```diff
@@ -14,13 +14,15 @@ L.AreaSelect = L.Class.extend({
 
     this._width = this.options.width;
     this._height = this.options.height;
+    this._bbox = this.options.bbox;
```
What is this needed for?

stephenkilbourn replied:
this was a stray from work on #388 .  I've removed it

m-mohr commented on src/components/map/leaflet-areaselect.js:
```diff
@@ -14,13 +14,15 @@ L.AreaSelect = L.Class.extend({
     minHorizontalSpacing: 30,
     minVerticalSpacing: 30,
     keepAspectRatio: false,
+    bbox: null,
```
What is this needed for?

stephenkilbourn replied:
this was a stray default from work on #388 .  I've removed it

m-mohr commented on src/components/SearchFilter.vue:
```diff
@@ -518,6 +534,11 @@ export default {
+    updateBBoxArray(entry, position) {
+      const bbox = this.query.bbox;
+      bbox[position] = Number(entry);
+      this.$set(this.query, 'bbox', bbox);
```
```suggestion
      this.setBBox(bbox);
```

m-mohr commented:
Thanks for the updates. Left a couple of comments, but otherwise looks good.

stephenkilbourn commented:
thanks for reviewing @m-mohr - comments addressed.  Things are working when I test locally against `https://earth-search.aws.element84.com/v1/`. Let me know if there are other issues.

stephenkilbourn commented:
I was not moving the map to center on the manually entered bounding box when a user went back to the map view. 

I added to commits with a few changes to address that bug and other potential ones:  
- With a little more testing around that, I realized a user could put also enter values for lat & lng in the Southwest corner that were larger than the NE corner values. I added some better validation there.

- There was also some shift in the exact values of bounding box on the map vs what was entered in the manual entry tab due to rounding. The map's resolution doesn't allow very precise drawings, but the leaflet function returns a float.  A 73.xxxxx1 in the manual entry would shift some when switching to the map tab and then back. I added a check to not get the new bounds on adding the area select if the bounds are already in state.  I also updated to round the lat & lng on the bounding box to an integer if the Map view is used to select a bounding box.

m-mohr commented:
Hey there,
sorry for the slow follow-up. I tested this again and it works better now, but the rounding seems to be flawed when smaller areas (e.g. a small pond or village) is selected. Then the rounding may lead to the following:

![grafik](https://github.com/radiantearth/stac-browser/assets/8262166/9170c5bc-f3bd-47ff-bb22-ad20bade1118)

luipir commented:
great to have this feature if aligned to use ol-stac

m-mohr commented:
You can simply rebase or merge the current changes into this branch here and resolve the potential merge conflicts. It probably needs quite a bit of an update to to the switch from Leaflet to OpenLayers and VueJS/Bootstrap. You can continue in this PR, no need to open a new one (unless you want to start fresh from scratch). Thank you! Let me know if you have questions.
